package/gluon-scheduled-domain-switch/luasrc/usr/bin/gluon-switch-domain

```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/lua
+
+local uci = require('simple-uci').cursor()
+local unistd = require 'posix.unistd'
+local site = require 'gluon.site'
+
+-- Returns true if given domain is present on device
+function domain_exists(domain_name)
+	return unistd.access("/lib/gluon/domains/"..domain_name..".json")
+end
+
+-- Returns true if node was offline long enough to perform domain switch
+function switch_after_min_reached()
+	if not unistd.access("/tmp/gluon_gw_unreach") then
+		return false
+	end
+
+	local current_epoch = os.time()
+	local switch_after_sec = site.domain_switch.switch_after_offline_mins() * 60
+
+	local f = io.open("/tmp/gluon_gw_unreach")
+	if not f then
+		return false
+	end
+	local offline_since = f:read('*a')
+	f:close()
+
+	local offline_time_sec = current_epoch - offline_since
+
+	if offline_time_sec > switch_after_sec then
```

Comment:
Could this result in a race condition right after boot, when no time was fetched via NTP yet?

For instance like:
1) Device booted, timestamp is equivalent to something like 1970
2) No gateways were discovered yet
3) /usr/sbin/gluon-check-gw-reachability writes a very low value to /tmp/gluon_gw_unreach
4) NTP updates the time a recent one.
5) /usr/sbin/gluon-check-gw-reachability is not called yet (or it is called but with no gateways found yet, or batctl erroring)
6) Now /usr/bin/gluon-switch-domain is called and switches immediately as it thinks there was unreachability for >40 years.

Maybe add some sanity check for any retrieved time? For instance, if any retrieved time is < 2018 then do not switch?

Reply:
The initial time after booting correlates to the time the image was built (and maybe the latest date in /etc, but i'm not sure on this one).

So a sanity check based on a minimal value will not help as all images will have been built in 2018.

A possible approach would be to compare the offline-time to the systems uptime, what do you think about that? This should also mitigate the possible race condition reading and writing this file.

Reply:
Using uptime instead of the Unix epoch for this issue is a great idea! I like that one.

I do not quite understand your last remark, how it fixes the read/write race condition though. The following could still happen, couldn't it?

1) Device booted, gets connected and sees some gateways
2) Uptime increases to n days. Still many days from the configured switching date away, though.
3) Gateway connectivity is lost somehow, for a short amount of time. But enough to make /usr/sbin/gluon-check-gw-reachability start writing its file. 
4) Now  /usr/bin/gluon-switch-domain is called while gluon-check-gw-reachability has only written the first (few) digit(s) of the current uptime so far. /usr/bin/gluon-switch-domain then compares this cut value to its current uptime and thinks ~n days of unreachability have passed.

Reply:
You are absolutely right, in this case the switch would still happen.

Reply:
Any sane date implementation will use stdio for its output, so short writes (up to 1KB in musl, I think 4KB is common in other libcs) happen atomically (with some caveats, but let's not go into detail too much here).

If you really want to be sure, you could write to a temporary file and then move it into place, but I don't consider this an actual issue.

Reply:
@blocktrron: Why not switching to store the uptime in /tmp/gluon_gw_unreach, too? Seems to me that this would simplify the checks a lot. Just compare the uptime in that file and the current uptime. No need to worry about NTP at all (at least for the switch_after_min_reached() check).

Reply:
@T-X great idea! I've implemented your suggestion, so we do not rely on the date and time anymore.